src/misc/makepattern.1
.TH MAKEPATTERN 1 "Feb 27 2014" "GMT 4.5.13 (SVN)" "Generic Mapping Tools"
.SH NAME
makepattern \- make GMT color pattern from b/w pattern or icon
.SH SYNOPSIS
\fBmakepattern\fP \fI1bit.ras\fP | \fIiconfile\fP \fB\-Cf\fP\fIr/g/b\fP \fB\-Cb\fP\fIr/g/b\fP
.SH DESCRIPTION
\fBmakepattern\fP reads either a Sun 1-bit rasterfile OR a Sun icon file.  It then accepts color choices for the foreground and
background pixels and colorizes the pattern, writing it as a 8-bit Sun rasterfile on stdout.
These patterns may then be used in \fBGMT\fP (3.1 or later) \fB\-Gp\fP settings or by \fBpsimage\fP.
.TP
\fI1bit.ras\fP | \fIiconfile\fP
Either a 1-bit Sun rasterfile (standard format, no RLE) or a Sun icon file (as used in \fBGMT\fP 3.0).
.TP
\fB\-Cf\fP
Sets the color for the foreground pixels (the ones) [black].
.TP
\fB\-Cb\fP
Sets the color for the background pixels (the zeros) [white].
.SH WIN32 fix
Since binary redirection under WIN32 does not work, you must specify the
output file with \fB\-G\fP\fInewfile.ras\fP.  This option is only available
under WIN32.
.SH EXAMPLES
.sp
To create a colorized red/blue 8-bit Sun rasterfile pattern from the old 3.0-style iconpattern
stored in file custom.icon, run
.sp
\fBmakepattern\fP custom.icon \fB\-Cf\fP red \fB\-Cb\fP blue > custom.ras
.sp
To create a green/blue 8-bit Sun rasterfile pattern from a 1-bit Sun raster called bits.ras, run
.sp
\fBmakepattern\fP bits.ras \fB\-Cf\fP green \fB\-Cb\fP blue > colorbits.ras
.SH "SEE ALSO"
.IR GMT (1),
.IR psimage (1)


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 12925 2014-02-20 22:02:24Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
201x-yy-zz	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---
2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.
2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		We consider the change in SDIST, PDIST to return degrees instead
				of km to be a bug as it makes it difficult to handle geodesics
				for which degrees make no sense.  Thus we have consolidated all
				the ?DIST operators to give km if geographic data are involved.
				Example 23 has been adjusted accordingly, as has several test
				scripts.  Please check your own scripts for SDIST.

2014-01-09	pwessel
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
2014-01-09	remko
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
	* grdmath.c:		LDIST and PDIST will again produce km when using geodesics.

2014-01-01	pwessel
	* x2sys/x2sys.c:	Mishandled the assignment of segment number for each record.
	* gmt_support.c:	Allow GMT_crossover to deal with global wrapped data.

2013-12-31	pwessel
	* gmt2kml.c:		Now using GMT_Put_Record instead of old stdout printing.
				This addresses issue #473.

2013-12-10	pwessel
	* grdvector.c:		Vector direction did not adjust when negative scales were
				used in -JX or -Jx.
	* psxy[z].c		---"---

2013-12-08	remko
	* psimage.c:		Now properly plots transparent images and makes -Gt work.

2013-11-26	pwessel
	* gmt_io.c:		GMT would crash on a blank line with no newline at end at all.
	* pstext.c:		pstext did not process GMT4-style input properly.

2013-11-26	jluis
	* potential/gravfft.c:	Let -E options default to 3.
				Let users select the -N+l (leave trend alone). That is, do not overwrite with default mode.
				Let the default detrend be remove MID point.
				Explicitly set -N+a in gravfft command to avoid failures due to coming changes in gravfft.
2013-11-25	pwessel
	* grdlandmask.c:	Failed to fill partial bins with no coastlines when the region
				crossed Greenwich [issue #451].

2013-11-21	pwessel
	* pscontour.c:		Let pscontour output contours to file [issue #448].
	* pscoast.c:		Prevent users from not giving proper projection [issue #449].

2013-11-19	jluis
	* mgd77/mgd77magref.c:	Needed to reset the t_col variable when constant altitude [issue #446].

2013-11-19	pwessel
	* gmt_sph.c:		Fix invalid read in ssrfpack_grid due to par array being of length 2 rather than 3.

2013-11-18	pwessel
	* gmt_grdio.c:		Place the old auto-detect of 4-byte binary float vs int under compatibility mode for backwardness.

2013-11-16	pwessel
	* grdvector.c:		Address shortcomings of geovectors when wrapped and clipped.

2013-11-15	pwessel
	* geovectors.c:		Fix geovector when shrinking is selected.

2013-11-13	pwessel
	* gmt_api.c:		Fix API issues pointed out by Ian.

2013-11-13	fwobbe
	* GMT.app:		Inherit user shell in GMT.app.

2013-11-12	pwessel
	* gmt_init.c:		grdmath could corrupt the gmt.history file [issue # 438].
 	
2013-11-10	pwessel
	* gmt_support.c:	Fix symbol-size issue in psxy -Sc [issue #435].
	
2013-11-09	pwessel
	* grdvector.c:		Did not convert direction to azimuths for geo vectors.

2013-11-08	pwessel
	* grdvector.c:		Fix grdvector -I problem [issue #434].
	* ps2raster.c:		Under Mavericks, strcpy would exit if the to and from pointers
-----------------------------------------------------------------------------------------
2013-11-05	pwessel
	GMT 5.1.0 finalized and released.
-----------------------------------------------------------------------------------------

2013-11-01	remko
	* example_10.sh:	Changed input to number of languages; multi-color columns

2013-10-31	pwessel
	* gmt_dcw.c:		Added option to specify code as =<continent> which will
				automatically be expanded to all the countries in that
				continent.  E.g., -F=AF+gred will paint Africa read.
				Fixed memory leaks.

2013-10-23	pwessel
	* gmtinfo.c:		The new name of minmax.c, the latter will be honored
				when GMT compatibility mode is 4.
	* grdmath.c:		Fixed porting bugs for CURV and D2DY2; added test ops.sh

2013-10-22	pwessel
	* minmax.c:		Added -D to adjust region found with -I
	* makecpt.c:		Better reporting of gaps or overlaps in CPTs

2013-10-05	pwessel
	* greenspline.c:	Needed -f in order to select -f0T input, plus it made assumptions
				about lon,lat despite not being selected.

2013-10-04	pwessel
	* gmt.history:		This replaces the old .gmtcommands file name.

2013-09-28	remko
	* gmt_init.c:		Moved origin of JD by one day (24 Nov instead of 25 Nov).
	* ConfigUserTemplate.cmake:	Added option GMT_DOCS_DEPEND_ON_GMT

2013-09-27	pwessel
	* share/cpt:		Added 10 new CPTs, mostly for bathymetry and topography.

2013-09-14	pwessel

src/misc/makepattern.c
/*--------------------------------------------------------------------
 *	$Id: makepattern.c 10173 2014-01-01 09:52:34Z pwessel $
 *
 *      Copyright (c) 1999-2014 by P. Wessel
 *      See LICENSE.TXT file for copying and redistribution conditions.
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; version 2 or any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      Contact info: www.soest.hawaii.edu/wessel
 *--------------------------------------------------------------------*/
/* makepattern is a utility for creating color imagepatterns to be used
 * with psimage or as pattern fill -Gi<size>/<image>.  It accepts either
 * an icon file OR a 1-bit standard Sun rasterfile and
 * assigns a foreground and background color where the bitimage is 1 or
 * 0, respectively, and writes out an 8-bit standard Sun rasterfile.
 *
 *--------------------------------------------------------------------*/
/*
  
   Author: 	P. Wessel
   Date:	19 MAR, 1998.
   version:	1.1
		1.2, 12-APR-1999 (DOS-compatible)

*/

#include "gmt.h"
#include "pslib.h"

/* Structure for Sun rasterfile */

int main (int argc, char **argv)
{
	GMT_LONG pad = FALSE, icon = FALSE;

	int i, j, k, kk, nx = 0, ny = 0, extra, mx, step, error = 0;

	unsigned char map[6], *rgb = NULL;
#ifdef _WIN32
	char *ofile = NULL;
#endif

	unsigned int p;
	unsigned short int *pattern = VNULL;

	char *fname = CNULL;

	struct GMT_FILL fg, bg;

	struct imageinfo h;

	unsigned short int *get_icon (char *file, int *nx, int *ny, int invert);

	FILE *fp = NULL, *fpo = NULL;

	GMT_init_fill (&fg, 0, 0, 0);		/* Default foreground is black */
	GMT_init_fill (&bg, 255, 255, 255);	/* Default background is white */

	argc = (int)GMT_begin (argc, argv);

	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			switch (argv[i][1]) {


				/* Common parameters */

				case '\0':
					error += (int)GMT_parse_common_options (argv[i], 0, 0, 0, 0);
					break;

				/* Supplemental parameters */

				case 'C':
					switch (argv[i][2]) {
						case 'f':
							if (GMT_getfill (&argv[i][3], &fg)) {
								GMT_fill_syntax ('C', " ");
								error++;
							}
							break;
						case 'b':
							if (GMT_getfill (&argv[i][3], &bg)) {
								GMT_fill_syntax ('C', " ");
								error++;
							}
							break;
						default:
							error++;
					}
					break;
#ifdef _WIN32
				case 'G':
					ofile = &argv[i][2];
					break;
#endif
				default:
					error = TRUE;
					GMT_default_error (argv[i][1]);
					break;
			}
		}
		else {
			fname = argv[i];
			if ((fp = fopen (fname, "rb")) == NULL) {
				fprintf(stderr,"%s: Cannot open file %s\n", GMT_program, fname);
				exit (EXIT_FAILURE);
			}
		}
	}

#ifdef _WIN32
	if (!ofile) {
		fprintf(stderr,"%s: No output file given\n", GMT_program);
		exit (EXIT_FAILURE);
	}
	if ((fpo = fopen (ofile, "wb")) == NULL) {
		fprintf(stderr,"%s: Cannot create file %s\n", GMT_program, ofile);
		exit (EXIT_FAILURE);
	}
#else
	fpo = stdout;
#endif
	if (argc == 1 || GMT_give_synopsis_and_exit) {
		fprintf (stderr, "makepattern %s - make color pattern from b/w pattern\n\n", GMT_VERSION);
#ifdef _WIN32
		fprintf(stderr,"usage:	makepattern 1bit.ras OR iconfile -Cf<rgb> -Cb<rgb> -Gimage.ras\n\n");
#else
		fprintf(stderr,"usage:	makepattern 1bit.ras OR iconfile -Cf<rgb> -Cb<rgb> > image.ras\n\n");
#endif

		if (GMT_give_synopsis_and_exit) exit (EXIT_FAILURE);

			fprintf(stderr,"\tGive the name of a 1-bit standard Sun rasterfile or iconfile\n");
		fprintf (stderr, "\n\tOPTIONS:\n");
		fprintf(stderr,"\t-Cf  Set foreground r/g/b color [0/0/0]\n");
		fprintf(stderr,"\t-Cb  Set background r/g/b color [255/255/255]\n");
#ifdef _WIN32
		fprintf(stderr,"\t-G  Name of output Sun rasterfile\n");
#endif
		exit (EXIT_FAILURE);
	}

	if (!fp) {
		fprintf(stderr,"%s: Specify either a 1-bit rasterfile or an iconfile\n", GMT_program);
		exit (EXIT_FAILURE);
	}
	else {	/* Got a 1-bit Sun rasterfile */

		if (ps_read_rasheader (fp, &h, 0, 7)) {
			fprintf (stderr, "%s: Trouble reading Sun rasterfile header!\n", GMT_program);
			exit (EXIT_FAILURE);
		}
		if (h.magic != RAS_MAGIC) {	/* Try to read as icon instead */
			fclose (fp);
			pattern = get_icon (fname, &nx, &ny, FALSE);
			icon = TRUE;
		}
		else {

			if (h.depth != 1) {
				fprintf (stderr, "%s: File is not a 1-bit Sun rasterfile!\n", GMT_program);
				exit (EXIT_FAILURE);
			}

			nx = h.width;
			ny = h.height;
			mx = h.length/2;
			pattern = (unsigned short int *) GMT_memory (VNULL, (size_t)mx, sizeof (unsigned short int), GMT_program);
			if (fread ((void *)pattern, sizeof (unsigned short int), (size_t)mx, fp) != (size_t)mx) {
				fprintf (stderr, "%s: Trouble reading 1-bit Sun rasterfile image!\n", GMT_program);
				exit (EXIT_FAILURE);
			}
			fclose (fp);
		}
	}

	/* Fill out the 2-entry color table with chosen back and foreground colors */

	map[0] = bg.rgb[0];	map[1] = fg.rgb[0];
	map[2] = bg.rgb[1];	map[3] = fg.rgb[1];
	map[4] = bg.rgb[2];	map[5] = fg.rgb[2];

	/* Allocated byte array for 8-bit image; each row must be even # of bytes, pad if necessary */

	mx = nx;
	if (nx%2) {
		pad = TRUE;
		mx++;	/* Must be even number of bytes */
	}
	rgb = (unsigned char *) GMT_memory (VNULL, (size_t)(mx * ny), sizeof (unsigned char), GMT_program);

	mx = nx / 16;			/* Number of full 16-bit shorts */
	extra = nx - mx * 16;		/* Remainder of bits in the last short */
	step = (extra) ? mx + 1 : mx;	/* Number of shorts per row */

	for (j = kk = 0; j < ny; j++) {		/* For each row in image */

		for (i = 0; i < mx; i++) {	/* For each chunk of full 16 bits */

			for (k = 0; k < 16; k++, kk++) {	/* Deal with each bit */

				p = (32768 >> k);
				rgb[kk] = (((unsigned int)(pattern[j*step+i]) & p) > 0);

			}
		}
		if (extra) {	/* Deal with remainder of bits in last short */
			i++;
			for (k = 0; k < extra; k++, kk++) {
				p = (32768 >> k);
				rgb[kk] = (((unsigned int)(pattern[j*step+i]) & p) > 0);
			}
		}
		if (pad) kk++;
	}

	/* Fill out header structure for rasterfile */

	mx = nx;
	if (mx%2) mx++;	/* Must be even number of bytes */

	h.magic = RAS_MAGIC;
	h.width = nx;
	h.height = ny;
	h.depth = 8;
	h.length = mx * ny;
	h.type  = RT_STANDARD;
	h.maptype = RMT_EQUAL_RGB;
	h.maplength = 6;

	if (ps_write_rasheader (fp, &h, 0, 7)) {
		fprintf (stderr, "%s: Trouble writing Sun rasterfile header!\n", GMT_program);
		exit (EXIT_FAILURE);
	}

	if (fwrite ((void *)map, sizeof (unsigned char), (size_t)h.maplength, fpo) != (size_t)h.maplength) {
		fprintf (stderr, "%s: Trouble writing Sun rasterfile colormap!\n", GMT_program);
		exit (EXIT_FAILURE);
	}

	if (fwrite ((void *)rgb, sizeof (unsigned char), (size_t)h.length, fpo) != (size_t)h.length) {
		fprintf (stderr, "%s: Trouble writing Sun rasterfile image!\n", GMT_program);
		exit (EXIT_FAILURE);
	}

#ifdef _WIN32
	fclose (fpo);
#endif

	GMT_free ((void *)rgb);
	if (icon)
		free ((void *)pattern);
	else
		GMT_free ((void *)pattern);

	GMT_end (argc, argv);

	exit (EXIT_SUCCESS);
}


unsigned short int *get_icon (char *file, int *nx, int *ny, int invert)
{
	int i, n_items, n_int, j, k, n_read, n_alloc, n_lines, last;
	unsigned int integer;
	unsigned short int *pattern;
	char t[8][7], line[80], width[12], height[12], *not_used = NULL;
	FILE *fp;

	if ((fp = fopen (file, "r")) == NULL) {
		fprintf (stderr, "%s: Cannot open file %s\n", GMT_program, file);
		return ((unsigned short int *)NULL);
	}

	not_used = fgets (line, 80, fp);
	line[strlen(line)-1] = 0;
	sscanf (&line[3], "%*[^,], %[^,], %[^,]", width, height);
	i = 0;	while (width[i] != '=') i++;
	*nx = atoi (&width[i+1]);
	i = 0;	while (height[i] != '=') i++;
	*ny = atoi (&height[i+1]);
	not_used = fgets (line, 80, fp);
	line[strlen(line)-1] = 0;
	n_items = (*nx) * (*ny) / 4;
	n_int = n_items / 4;
	n_alloc = (int)ceil (n_int / 8.0) * n_int;
	pattern = (unsigned short int *) malloc ((size_t)n_alloc);
	n_lines = n_items / 32;
	for (i = k = 0; i < n_lines; i++) {
		not_used = fgets (line, 80, fp);
		line[strlen(line)-1] = 0;
		last = (i == (n_lines - 1));
		if (last) {
			n_read = sscanf (&line[1], "%[^,],%[^,],%[^,],%[^,],%[^,],%[^,],%[^,],%[^,]", t[0], t[1], t[2], t[3], t[4], t[5], t[6], t[7]);
			if (n_read == 4) {
				sscanf (&line[1], "%[^,], %[^,], %[^,], %[^,]", t[0], t[1], t[2], t[3]);	/* New style icon file */
				not_used = fgets (line, 80, fp);
				line[strlen(line)-1] = 0;
				sscanf (&line[1], " %[^,], %[^,], %[^,], %[^,]", t[4], t[5], t[6], t[7]);	/* New style icon file */
			}
		}
		else {
			n_read = sscanf (&line[1], "%[^,],%[^,],%[^,],%[^,],%[^,],%[^,],%[^,],%[^,],", t[0], t[1], t[2], t[3], t[4], t[5], t[6], t[7]);
			if (n_read == 4) {
				sscanf (&line[1], "%[^,], %[^,], %[^,], %[^,],", t[0], t[1], t[2], t[3]);	/* New style icon file */
				not_used = fgets (line, 80, fp);
				line[strlen(line)-1] = 0;
				sscanf (&line[1], " %[^,], %[^,], %[^,], %[^,],", t[4], t[5], t[6], t[7]);	/* New style icon file */
			}
		}
		for (j = 0; j < 8; j++, k++) {
			sscanf (&t[j][2], "%x", &integer);
			pattern[k] = integer;
		}
	}
	fclose (fp);
	if (invert) for (k = 0; k < n_int; k++) pattern[k] = ~pattern[k];
	return (pattern);
}
